Add command 'antctl upgrade api-storage' in antctl (#5198)
antrea-io/antrea: docs/antctl.md

worker2 172.18.0.3 Alive 
worker3 172.18.0.2 Dead
```

### Upgrade existing objects of CRDs

> antctl supports upgrading existing objects of Antrea CRDs to the latest version. The related sub-commands should be run

Reviewer: Wrap the lines at 80

Author: Done

worker2 172.18.0.3 Alive 
worker3 172.18.0.2 Dead
```

### Upgrade existing resources

> Antctl supports upgrading existing resources of CRDs managed by Antrea to the latest version. The related sub commands

Reviewer 2: You may highlight that the command will only handle CRDs with suffix `crd.antrea.io`:
```suggestion
Antctl supports upgrading existing resources of CRDs with suffix `crd.antrea.io` to the latest version. The related sub commands
```

Reviewer: Maybe simply say: antctl supports upgrading resources of Antrea CRDs to the latest version

> Antctl supports upgrading existing resources of CRDs managed by Antrea to the latest version. The related sub commands

Reviewer: Antctl -> antctl

Could you also change other "Antctl" in this doc?

Reviewer: sub commands -> sub-commands

Author: Done

should be run out-of-cluster.

> This command is used to upgrade all existing persisted resources to the latest API version:

Reviewer 2: ```suggestion
This command upgrades all existing persisted resources to the latest API version:
```

Reviewer: all existing persisted resources -> resources of Antrea CRDs

Author: Done

```bash
antctl upgrade api-storage
```

> This command is used to upgrade persisted resources for Antrea CRD `antreaagentinfos.crd.antrea.io` to the latest API

Reviewer 2: ditto

> This command is used to upgrade persisted resources for Antrea CRD `antreaagentinfos.crd.antrea.io` to the latest API

Reviewer: persisted resources for Antrea CRD `antreaagentinfos.crd.antrea.io` ->  AntreaAgentInfo resources

Author: Done.

version:

```bash
antctl upgrade api-storage --crd-list=antreaagentinfos.crd.antrea.io
```

> This command is used to upgrade persisted resources for Antrea CRD `egresses.crd.antrea.io` and `groups.crd.antrea.io`

Reviewer 2: ditto

Reviewer: Egress and Group resources

Author: Done

worker2 172.18.0.3 Alive 
worker3 172.18.0.2 Dead
```

### Upgrade existing objects of CRDs

> antctl supports upgrading existing objects of Antrea CRDs to the latest version.

Reviewer 3: ```suggestion
antctl supports upgrading existing objects of Antrea CRDs to the storage version.
```

Author: Should we change all `the latest version` to `the storage version`? Generally, the `storage version` should be the latest version. @tnqn

Author: Done

> The related sub-commands should be run out-of-cluster.

Reviewer 3: We should mention the required permissions. It includes the get, list, and update of each CRD and its objects.
And make sure proper error is printed when some permissions are missing.

Author: Will add.

Reviewer 3: Running outside the cluster doesn't mean the kubeconfig file has admin priviledge. Normally we should add the required permissions of this command to ClusterRole antctl as the existence of the ClusterRole implies it contains all permissions its commands require. However, the required permissions of `upgrade api-storage` could make antctl a super user, while its privilege is currently quite limited, the only create permission it get is creating supportbundles.

So a conservative approach is to document clearly what permissions are required for the upgrade command, then users know what they need to do when they meet encounter issue.

I don't know how `antctl mc` guides users to get required permissions. But for other commands, we have added all permissions to ClusterRole antctl. For `upgrage api-storage`, we need at least document it.

cc @jianjuns @antoninbas

Author: Got that, I'll add the required permissions in doc and give some error prints for lacking permissions.

Reviewer 2: I think there is no document or permissions added to ClusterRole when we add the `antctl mc`, I will create an issue to track this first.

Author: Done

worker2 172.18.0.3 Alive 
worker3 172.18.0.2 Dead
```

### Upgrade existing objects of CRDs

antctl supports upgrading existing objects of Antrea CRDs to the storage version.
The related sub-commands should be run out-of-cluster. Please ensure that the
kubeconfig file used by antctl has the necessary permissions. The required permissions

> are listed in the following example ClusterRole.

Reviewer 2: ```suggestion
are listed in the following sample ClusterRole.
```

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:

>   name: role-for-antctl

Reviewer 2: The ClusterRole name in Antrea is antctl, so you may use `antctl` to replace `role-for-antctl`.

rules:
  - apiGroups:
      - apiextensions.k8s.io
    resources:
      - customresourcedefinitions
    verbs:
      - get 
      - list
  - apiGroups:
      - apiextensions.k8s.io
    resources: 
      - customresourcedefinitions/status
    verbs:
      - update
  - apiGroups: 
      - crd.antrea.io
    resources:
      - "*"
    verbs: 
      - get
      - list 
      - update
```

This command performs a dry-run to upgrade all existing objects of Antrea CRDs to
the storage version:

```bash
antctl upgrade api-storage --dry-run
```

This command upgrades all existing objects of Antrea CRDs to the storage version:

```bash
antctl upgrade api-storage
```

This command upgrades existing AntreaAgentInfo objects to the storage version:

```bash
antctl upgrade api-storage --crds=antreaagentinfos.crd.antrea.io
```

This command upgrades existing Egress and Group objects to the storage version:

```bash
antctl upgrade api-storage --crds=egresses.crd.antrea.io,groups.crd.antrea.io
```

If you encounter any errors related to permissions while running the commands, double-check
the permissions of the kubeconfig used by antctl. Ensure that the ClusterRole has the

> required permissions. Example of errors related to insufficient permissions:

Reviewer 2: ```suggestion
required permissions. The following sample errors are caused by insufficient permissions:
```
